common/workunit/workunit.cpp

```diff
@@ -6359,6 +6429,83 @@ unsigned CLocalWorkUnit::queryFileUsage(const char *fileName) const
     return p->getPropInt(path.str());
 }
 
+IConstWUFileUsageIterator * CLocalWorkUnit::getFieldUsage() const
+{
+    CriticalBlock block(crit);
+    IPropertyTree* fieldUsageTree = p->queryPropTree("usedsources");
+
+    if (!fieldUsageTree)
+        return NULL;
+
+    IPropertyTreeIterator* iter = fieldUsageTree->getElements("*");
+    return new CConstWUFileUsageIterator(iter);
+}
+
+bool CLocalWorkUnit::getFieldUsageArray(StringArray & filenames, StringArray & columnnames, const char * clusterName) const
+{
+    bool scopeLoaded = false;
+    SCMStringBuffer defaultScope;
+
+    Owned<IConstWUFileUsageIterator> files = getFieldUsage();
+
+    if (!files)
+        return false; // this query was not compiled with recordFieldUsage option.
+
+    ForEach(*files)
+    {    
+        Owned<IConstWUFileUsage> file = files->get();
+
+        SCMStringBuffer filename;
+        file->getName(filename);
+
+        // filename cannot be empty
+        if (filename.length() < 3)
+            throw MakeStringException(WUERR_InvalidFieldUsage, "Invalid FieldUsage found in WU. Cannot enforce view security.");
+
+        // Filenames in field usage are surrounded with single quotes ('filename').
```

Comment:
I made a fix so that when a filename is not surrounded by a single quote, it is not treated as a resolved filename. When a query contains any unresolved filename, it means a query may potentially access ANY files, so it will be denied unconditionally (in terms of view security).